#!/usr/bin/env python3

#=======================================================================
# Authors: Ben Woodcroft
#
# Unit tests.
#
# Copyright
#
# This is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License.
# If not, see <http://www.gnu.org/licenses/>.
#=======================================================================

import unittest
import os.path
import sys
import extern
import re
import tempfile
import json
import math
import polars as pl

import pytest

sys.path = [os.path.join(os.path.dirname(os.path.realpath(__file__)),'..')]+sys.path

from bird_tool_utils import in_tempdir

from singlem.metapackage import Metapackage

path_to_script = 'singlem'
path_to_data = os.path.join(os.path.dirname(os.path.realpath(__file__)),'data')

singlem_base_directory = os.path.join(os.path.dirname(__file__), '..')
# TODO: Once GTDBtk can be included in conda env (as of diamond 2.1.7 likely), remove the added PATH entry
singlem_bin_directory = os.path.join(singlem_base_directory, 'bin')
# run_supplement = f"GTDBTK_DATA_PATH=/work/microbiome/db/gtdb/gtdb_release207_v2 PATH=$PATH:{singlem_bin_directory} {path_to_script} supplement"
run_supplement = f"singlem supplement"
singlem = f"{singlem_bin_directory}/singlem"

@pytest.mark.expensive
class Tests(unittest.TestCase):
    '''Tests which require a full metapackage installation to run, so can't be run through GitHub actions.'''
    maxDiff = None
    otu_table_headers = str.split('gene sample sequence num_hits coverage taxonomy')

    # Used in test_outside_ci.py too
    def assert_equal_taxonomic_profile(self, observed_string, expected_array_or_string):
        observed_array = list([line.split("\t") for line in observed_string.split("\n")])

        r = re.compile(r'  +')
        if isinstance(expected_array_or_string, str):
            expected_array = list(expected_array_or_string.split("\n"))
            expected_array = [r.sub("\t", line) for line in expected_array]
            expected_array = [line.split("\t") for line in expected_array]
        else:
            expected_array = expected_array_or_string

        if expected_array[-1] != ['']:
            expected_array.append([''])

        # make sure headers are OK
        self.assertEqual(expected_array[0], observed_array[0])

        # sort the rest of the table and compare that
        self.assertEqual(sorted(expected_array[1:]), sorted(observed_array[1:]))

    def assertEqualOtuTable(self, expected_array, observed_string):
        observed_array = list([line.split("\t") for line in observed_string.split("\n")])
        if expected_array[-1] != ['']:
            expected_array.append([''])

        # make sure headers are OK
        self.assertEqual(expected_array[0], observed_array[0])

        # sort the rest of the table and compare that
        self.assertEqual(sorted(expected_array[1:]), sorted(observed_array[1:]))

    def test_condense_cli(self):
        '''Test the condense CLI.'''
        with in_tempdir():
            cmd1 = "{} pipe --genome-fasta-files {}/methanobacteria/transcriptomes/GB_GCA_000309865.1_protein.fna --archive-otu-table {}".format(
                path_to_script, path_to_data, 'mbac.json')
            extern.run(cmd1)
            cmd = "{} condense --input-archive-otu-table mbac.json -p /dev/stdout".format(path_to_script)
            observed = extern.run(cmd)
            expected = '\n'.join(
                ['sample  coverage        taxonomy',
                'GB_GCA_000309865.1_protein      1.12    Root; d__Archaea; p__Methanobacteriota; c__Methanobacteria; o__Methanobacteriales; f__Methanobacteriaceae; g__Methanobacterium; s__Methanobacterium sp000309865\n'])
            self.assert_equal_taxonomic_profile(observed, expected)

    # This test takes a long time - like 1+ hours.
    def test_supplement_with_extra_taxon_genome_lengths(self):
        with in_tempdir():
            cmd = f"{run_supplement} --no-dereplication --skip-taxonomy-check --hmmsearch-evalue 1e-5 --no-quality-filter --new-genome-fasta-files {path_to_data}/supplement/GCA_011373445.1_genomic.mutated93_ms.manually_added_nongaps.fna --output-metapackage out.smpkg --new-fully-defined-taxonomies {path_to_data}/supplement/GCA_011373445.1_genomic.mutated93_ms.manually_added_nongaps.fna.taxonomy --checkm2-quality-file {path_to_data}/supplement/checkm2.output/quality_report.tsv"
            # print(cmd)
            extern.run(cmd)

            cmd2 = f'{path_to_script} pipe --genome-fasta-files {path_to_data}/supplement/GCA_011373445.1_genomic.mutated93_ms.manually_added_nongaps.fna --metapackage out.smpkg/ --otu-table /dev/stdout'
            output = extern.run(cmd2)
            expected = [
                "\t".join(self.otu_table_headers),
                '4.11.22seqs	GCA_011373445.1_genomic.mutated93_ms.manually_added_nongaps	CTTAAAAAGAAACTAAAAGGTGCCGGCGCTCACATGAGGGTTCTAAAAAACACTCTAATT	1	1.18	Root; d__Archaea; p__Thermoproteota; c__Bathyarchaeia; o__B26-1; f__UBA233; g__DRVV01; s__NEW_SPECIES'
            ]

            new_mpkg = Metapackage.acquire('out.smpkg')
            new_lengths = pl.DataFrame(new_mpkg.taxon_genome_lengths())
            new_species_entry = new_lengths.filter(pl.col('rank')=='s__NEW_SPECIES')
            self.assertEqual(new_species_entry.shape[0], 1)
            self.assertEqual(new_species_entry['genome_size'][0], 3937114.9507050873)

            old_mpkg = Metapackage.acquire(os.environ['SINGLEM_METAPACKAGE_PATH'])
            old_taxon_lengths = pl.DataFrame(old_mpkg.taxon_genome_lengths())
            self.assertEqual(len(old_taxon_lengths.filter(pl.col('rank')=='s__NEW_SPECIES')), 0)
            self.assertEqual(len(old_taxon_lengths)+1, len(new_lengths))

            self.assertAlmostEqual(
                old_taxon_lengths.filter(pl.col('rank')=='d__Bacteria')['genome_size'][0],
                new_lengths.filter(pl.col('rank')=='d__Bacteria')['genome_size'][0],
                places=3) 

            self.assertTrue(
                old_taxon_lengths.filter(pl.col('rank')=='d__Archaea')['genome_size'][0] < new_lengths.filter(pl.col('rank')=='d__Archaea')['genome_size'][0]) 
            self.assertTrue(
                new_lengths.filter(pl.col('rank')=='d__Archaea')['genome_size'][0] < 1.8e6)

    # # This test takes a long time - like 1+ hours. Currently commented out because the test code was not ever implemented properly.
    # def test_supplement_nothing_precalculated(self):
    #     with in_tempdir():
    #         cmd = f"{run_supplement} --hmmsearch-evalue 1e-5 --dereplicate-with-galah --no-quality-filter --new-genome-fasta-files {path_to_data}/supplement/GCA_011373445.1_genomic.mutated93_ms.manually_added_nongaps.fna --output-metapackage ~/git/singlem/test/tmp/test_supplement_nothing_precalculated.smpkg --checkm2-quality-file {path_to_data}/supplement/checkm2.output/quality_report.tsv --threads 8 &> ~/git/singlem/test/tmp/test_supplement_nothing_precalculated.log"
    #         # print(cmd)
    #         extern.run(cmd)

    #         cmd2 = f'{path_to_script} pipe --genome-fasta-files {path_to_data}/supplement/GCA_011373445.1_genomic.mutated93_ms.manually_added_nongaps.fna --metapackage out.smpkg/ --otu-table /dev/stdout'
    #         output = extern.run(cmd2)
    #         expected = [
    #             "\t".join(self.otu_table_headers),
    #             '4.11.22seqs	GCA_011373445.1_genomic.mutated93_ms.manually_added_nongaps	CTTAAAAAGAAACTAAAAGGTGCCGGCGCTCACATGAGGGTTCTAAAAAACACTCTAATT	1	1.18	Root; d__Archaea; p__Thermoproteota; c__Bathyarchaeia; o__B26-1; f__UBA233; g__DRVV01; s__NEW_SPECIES'
    #         ]

    #         new_mpkg = Metapackage.acquire('out.smpkg')
    #         new_lengths = pl.DataFrame(new_mpkg.taxon_genome_lengths())
    #         new_species_entry = new_lengths.filter(pl.col('rank')=='s__NEW_SPECIES')
    #         self.assertEqual(new_species_entry.shape[0], 1)
    #         self.assertEqual(new_species_entry['genome_size'][0], 3937114.9507050873)

    #         old_mpkg = Metapackage.acquire(os.environ['SINGLEM_METAPACKAGE_PATH'])
    #         old_taxon_lengths = pl.DataFrame(old_mpkg.taxon_genome_lengths())
    #         self.assertEqual(len(old_taxon_lengths.filter(pl.col('rank')=='s__NEW_SPECIES')), 0)
    #         self.assertEqual(len(old_taxon_lengths)+1, len(new_lengths))

    #         self.assertAlmostEqual(
    #             old_taxon_lengths.filter(pl.col('rank')=='d__Bacteria')['genome_size'][0],
    #             new_lengths.filter(pl.col('rank')=='d__Bacteria')['genome_size'][0],
    #             places=3) 

    #         self.assertTrue(
    #             old_taxon_lengths.filter(pl.col('rank')=='d__Archaea')['genome_size'][0] < new_lengths.filter(pl.col('rank')=='d__Archaea')['genome_size'][0]) 
    #         self.assertTrue(
    #             new_lengths.filter(pl.col('rank')=='d__Archaea')['genome_size'][0] < 1.8e6)

    def _assert_archive_otu_tables_equal(self, expected, observed):
        '''Compare two archive OTU table dicts, allowing minor floating-point differences in coverage values.'''
        self.assertEqual(expected['version'], observed['version'])
        self.assertEqual(expected['fields'], observed['fields'])
        self.assertEqual(expected['alignment_hmm_sha256s'], observed['alignment_hmm_sha256s'])
        self.assertEqual(expected['singlem_package_sha256s'], observed['singlem_package_sha256s'])
        coverage_idx = expected['fields'].index('coverage')
        def otu_key(otu):
            return (otu[0], otu[1], otu[2])
        exp_otus = sorted(expected['otus'], key=otu_key)
        obs_otus = sorted(observed['otus'], key=otu_key)
        self.assertEqual(len(exp_otus), len(obs_otus))
        for exp_otu, obs_otu in zip(exp_otus, obs_otus):
            for i, (e, o) in enumerate(zip(exp_otu, obs_otu)):
                if i == coverage_idx:
                    self.assertTrue(math.isclose(e, o, rel_tol=1e-9),
                        f"Coverage mismatch for {exp_otu[0]}/{exp_otu[1]}: {e} vs {o}")
                else:
                    self.assertEqual(e, o,
                        f"Field {expected['fields'][i]} mismatch for {exp_otu[0]}/{exp_otu[1]}")

    def test_sra_pipe_standard(self):
        with tempfile.NamedTemporaryFile(suffix='.json') as tf:
            cmd = "{} pipe --threads 4 --sra-files {}/SRR8653040.sra --no-assign-taxonomy --archive-otu-table {}".format(
                path_to_script, path_to_data, tf.name)
            extern.run(cmd)
            self._assert_archive_otu_tables_equal(
                json.load(open(os.path.join(path_to_data, 'SRR8653040.json'))),
                json.load(open(tf.name)))

    def test_sra_pipe_chunk(self):
        # root@a1d31a133bde:/# kingfisher extract --sra /tmp/SRR8653040.sra --stdout --unsorted -f fasta |wc -l
        # 848128
        #
        # => so 2nd 200,000 chunk is fine.
        with tempfile.NamedTemporaryFile(suffix='.json') as tf:
            cmd = "{} pipe --threads 4 --sra-files {}/SRR8653040.sra --no-assign-taxonomy --archive-otu-table {} --read-chunk-size 200000 --read-chunk-num 2".format(
                path_to_script, path_to_data, tf.name)
            extern.run(cmd)
            self._assert_archive_otu_tables_equal(
                json.load(open(os.path.join(path_to_data, 'SRR8653040.json'))),
                json.load(open(tf.name)))

    def test_rare_eif2_issue(self):
        # Here, insert size is small, so forward and reverse reads are matched,
        # but some differently because the insert means they have different
        # assignment procedures. See https://github.com/wwood/singlem/issues/265
        bad_seqs1 = """
>SRR27700371.281145 A01909:184:H2CNHDSX7:1:1106:1696:19398 length=151
GGTGCATTCGTTACTATCCTTCCAGGTAAAGATGGTCTAGTACACATCTCTCAAATCGCTGACAGGCGCGTTGAGAAAGTGTCTGACTACCTAACTGAAGGTCAAGAAGTACCTGTTAAGGTTCTTTAAATTGACCGTCAAGGCCGTGTAC
>SRR27700371.1403435 A01909:184:H2CNHDSX7:1:1128:12427:6355 length=151
GGTGCATTCGTTACTATCCTTCCAGGTAAAGATGGTCTAGTACACATCTCTCAAATCGCTGATAAGCGCGTTGAGAAAGTGTCTGACTACCTAACTGAAGGTCAAGAAGTACCAGTTAAGGTTCTTGAAATTGACCGTCAAGGCCGTGTAC
>SRR27700371.2581598 A01909:184:H2CNHDSX7:1:1150:32199:25285 length=151
ACAATGAAAATCAGCTCTGATAAGATCAAAGATGTTATCGGTAAAGGCGGCGCAGTTATCCGTGCTCTTTGTGAAGAAACAGGTACTACAATCGAAATCGAAGACGATGGCACAATCAAGATTGCTGCTACTGAAGGCGCAGCTGCTAAAG
>SRR27700371.2829642 A01909:184:H2CNHDSX7:1:1155:14326:16188 length=151
CTGCACTTCAAATGGACATCAAGATCGAAGGTATCACTAAAGAGATCATGCAAATTGCTCTTAACCAAGCGCAAGGTGCACGTAAGCACATCCTTTCTGTTATGGATGAAGCTATCTCAGGTGCTCGTGAAGATATCTCTGAATTCGCTCC
>SRR27700371.3755246 A01909:184:H2CNHDSX7:1:1173:4851:3161 length=151
ACATACCCGCGGTTTGAAACCATTTCACAAATAATTACTCCTGTTATGACCGCAATAAAAAAGGGGAGCGATATGCTCCCCTTTATAACATGCTTTTTTACCAATGGGTTAGAAATCACTAAAAAGCGATAGTCCCATTAATAGGCTAAGA
>SRR27700371.3967214 A01909:184:H2CNHDSX7:1:1177:29559:4335 length=151
TTCAAACAGCATCTTAGCTCGAACTTCATCGTTCAAATCCGGGCGCTGCAATAATTGACTAAGACGTGCAATTTGAACTTCTTGCTGAACGCTTGGTTGCAGCGGCACAGCCATTGGTGGATAAACCCAGCGTGAGGTGTTATCTGATGTT
>SRR27700371.3969379 A01909:184:H2CNHDSX7:1:1177:29044:5822 length=151
TTCAAACAGCATCTTAGCTCGAACTTCATCGTTCAAATCCGGGCGCTGCAATAATTGACTAAGACGTGCAATTTGAACTTCTTGCTGAACGCTTGGTTGCAGCGGCACAGCCATTGGTGGATAAACCCAGCGTGAGGTGTTATCTGATGTT
>SRR27700371.7890137 A01909:184:H2CNHDSX7:1:1273:15320:26396 length=151
TACACGGCCTTGACGGTCAATTTCAAGAACCTTAACAGGTACTTCTTGACCTTCAGTTAGGTAGTCAGACACTTTCTCAACGCGCTTGTCAGCGATTTGAGAGATGTGTACTAGACCATCTTTACCTGGAAGGATAGTAACGAATGCACCG
>SRR27700371.9410083 A01909:184:H2CNHDSX7:1:1324:12246:24831 length=151
GGTGCATTCGTTACTATCCTTCCAGGTAAAGATGGTCTAGTACACATCTCTCAAATCGCTGATAAGCGCGTTGAGAAAGTGTCTGACTACCTAACTGAAGGTCAAGAAGTACCAGTTAAGGTTCTTGAAATTGACCGTCAAGGCCGTGTAC
>SRR27700371.9594429 A01909:184:H2CNHDSX7:1:1328:26458:6605 length=151
TACACGGCCTTGACGGTCAATTTCAAGAACCTTAACTGGTACTTCTTGACCTTCAGTTAGGTAGTCAGACACTTTCTCAACGCGCTTATCAGCGATTTGAGAGATGTGTACTAGACCATCTTTACCTGGAAGGATAGTAACGAATGCACCG
>SRR27700371.13011471 A01909:184:H2CNHDSX7:1:1414:29577:33113 length=151
CGGTGCATTCGTTACTATCCTTCCAGGTAAAGATGGTCTAGTACACATCTCTCAAATCGCTGATAAGCGCGTTGAGAAAGTGTCTGACTACCTGACTGAAGGTCAAGAAGTACCAGTTAAGGTTCTTGAAATTGACCGTCAAGGCCGTGTA
>SRR27700371.13013700 A01909:184:H2CNHDSX7:1:1414:30852:34601 length=151
CGGTGCATTCGTTACTATCCTTCCAGGTAAAGATGGTCTAGTACACATCTCTCAAATCGCTGATAAGCGCGTTGAGAAAGTGTCTGACTACCTGACTGAAGGTCAAGAAGTACCAGTTAAGGTTCTTGAAATTGACCGTCAAGGCCGTGTA
>SRR27700371.14045650 A01909:184:H2CNHDSX7:1:1434:15962:20055 length=151
TACACGGCCTTGACGGTCAATTTCAAGAACCTTAACTGGTACTTCTTGACCTTCAGTTAGGTAGTCAGACACTTTCTCAACGCGCTTATCAGCGATTTGAGAGATGTGTACTAGACCATCTTTACCTGGAAGGATAGTAACGAATGCACCG
>SRR27700371.14048634 A01909:184:H2CNHDSX7:1:1434:14977:22106 length=151
TACACGGCCTTGACGGTCAATTTCAAGAACCTTAACTGGTACTTCTTGACCTTCAGTTAGGTAGTCAGACACTTTCTCAACGCGCTTATCAGCGATTTGAGAGATGTGTACTAGACCATCTTTACCTGGAAGGATAGTAACGAATGCACCG
>SRR27700371.14048653 A01909:184:H2CNHDSX7:1:1434:15022:22122 length=151
TACACGGCCTTGACGGTCAATTTCAAGAACCTTAACTGGTACTTCTTGACCTTCAGTTAGGTAGTCAGACACTTTCTCAACGCGCTTATCAGCGATTTGAGAGATGTGTACTAGACCATCTTTACCTGGAAGGATAGTAACGAATGCACCG
>SRR27700371.15916112 A01909:184:H2CNHDSX7:1:1469:7374:30107 length=151
TACACGGCCTTGACGGTCAATTTCAAGAACCTTAACTGGTACTTCTTGACCTTCAGTTAGGTAGTCAGACACTTTCTCAACGCGCTTATCAGCGATTTGAGAGATGTGTACTAGACCATCTTTACCTGGAAGGATAGTAACGAATGCACCG
>SRR27700371.17955650 A01909:184:H2CNHDSX7:1:1530:16170:23641 length=151
GTACACGGCCTTGACGGTCAATTTCAAGAACCTTAACTGGTACTTCTTGACCTTCAGTTAGGTAGTCAGACACTTTCTCAACGCGCTTATCAGCGATTTGAGAGATGTGTACTAGACCATCTTTACCTGGAAGGATAGTAACGAATGCACC
>SRR27700371.18351113 A01909:184:H2CNHDSX7:1:1538:1090:5384 length=151
GATAAACCCAGCGTGAGGTGTTATCTGATGTTGTCGCACAACCTGTCAGTACAAGCAGTAAACACATACTCGCGGTTTGAAACCATTTCACAAATAATTACTCCTGTTATGACCGCAATAAAAAAGGGGAGCGATATGCTCCCCTTTATAA
>SRR27700371.18716393 A01909:184:H2CNHDSX7:1:1545:3287:2832 length=151
GCCTTCAGCTGGCGTTTCAACTGCTTCTTTCATGCTTAGACGTACACGGCCTTGACGGTCAATTTCAAGAACCTTAACTGGTACTTCTTGACCTTCAGTTAGGTAGTCAGACACTTTCTCAACGCGCTTATCAGCGATTTGAGAGATGTGT
>SRR27700371.19217471 A01909:184:H2CNHDSX7:1:1554:1633:21637 length=151
ACACGGCCTTGACGGTCAATTTCAAGAACCTTAACTGGTACTTCTTGACCTTCAGTTAGGTAGTCAGACACTTTCTCAACGCGCTTATCAGCGATTTGAGAGATGTGTACTAGACCATCTTTACCTGGAAGGATAGTAACGAATGCACCGA
>SRR27700371.21639151 A01909:184:H2CNHDSX7:1:1622:12861:33191 length=151
ATGGATGAAGCTATCTCAGGTGCTCGTGAAGATATCTCTGAATTCGCTCCACGTATCCACACAATGAAAATCAGCTCTGATAAGATCAAAGATGTTATCGGTAAAGGCGGCGCAGTTATCCGTGCTCTTTGTGAAGAAACAGGTACTACAA
>SRR27700371.27341237 A01909:184:H2CNHDSX7:1:2155:3721:25942 length=151
TACACGGCCTTGACGGTCAATTTCAAGAACCTTAACTGGTACTTCTTGACCTTCAGTTAGGTAGTCAGACACTTTCTCAACGCGCTTATCAGCGATTTGAGAGATGTGTACTAGACCATCTTTACCTGGAAGGATAGTAACGAATGCACCG
>SRR27700371.29912070 A01909:184:H2CNHDSX7:1:2228:22453:10254 length=151
ATACTCGCGGTTTGAAACCATTTCACAAATAATTACTCCTGTTATGACCGCAATAAAAAAGGGGAGCGATATGCTCCCCTTTATAACATGCTTTTTTACCAATGGGTTAGAAATCACTAAAAAGCGATATTCCCATCAATACGCTAAGAAT
>SRR27700371.33707918 A01909:184:H2CNHDSX7:1:2324:15347:22435 length=151
GGTGCATTCGTTACTATCCTTCCAGGTAAAGATGGTCTAGTACACATCTCTCAAATCGCTGATAAGCGCGTTGAGAAAGTGTCTGACTACCTAACTGAAGGTCAAGAAGTACCAGTTAAGGTTCTTGAAATTGACCGTCAAGGCCGTGTAC
>SRR27700371.33938987 A01909:184:H2CNHDSX7:1:2329:2392:3286 length=151
GGTGCATTCGTTACTATCCTTCCAGGTAAAGATGGTCTAGTACACATCTCTCAAATCGCTGATAAGCGCGTTGAGAAAGTGTCTGACTACCTAACTGAAGGTCAAGAAGTACCAGTTAAGGTTCTTGAAATTGACCGTCATGGCCGTGTAC
>SRR27700371.33939283 A01909:184:H2CNHDSX7:1:2329:2582:3490 length=151
GGTGCATTCGTTACTATCCTTCCAGGTAAAGATGGTCTAGTACACATCTCTCAAATCGCTGATAAGCGCGTTGAGAAAGTGTCTGACTACCTAACTGAAGGTCAAGAAGTACCAGTTAAGGTTCTTGAAATTGACCGTCAAGGCCGTGTAC
>SRR27700371.33939842 A01909:184:H2CNHDSX7:1:2329:2501:3881 length=151
GGTGCATTCGTTACTATCCTTCCAGGTAAAGATGGTCTAGTACACATCTCTCAAATCGCTGATAAGCGCGTTGAGAAAGTGTCTGACTACCTAACTGAAGGTCAAGAAGTACCAGTTAAGGTTCTTGAAATTGACCGTCAGGGCCGTGTAC
>SRR27700371.34166394 A01909:184:H2CNHDSX7:1:2333:29134:18724 length=151
CGGTGCATTCGTTACTATCCTTCCAGGTAAAGATGGTCTAGTACACATCTCTCAAATCGCTGATAAGCGCGTTGAGAAAGTGTCTGACTACCTAACTGAAGGTCAAGAAGTACCAGTTAAGGTTCTTGAAATTGACCGTCAAGGCCGTGTA
>SRR27700371.34173398 A01909:184:H2CNHDSX7:1:2333:31412:23703 length=151
CGGTGCATTCGTTACTATCCTTCCAGGTAAAGATGGTCTAGTACACATCTCTCAAATCGCTGATAAGCGCGTTGAGAAAGTGTCTGACTACCTAACTGAAGGTCAAGAAGTACCAGTTAAGGTTCTTGAAATTGACCGTCAAGGCCGTGTA
>SRR27700371.34174518 A01909:184:H2CNHDSX7:1:2333:31647:24518 length=151
CGGTGCATTCGTTACTATCCTTCCAGGTAAAGATGGTCTAGTACACATCTCTCAAATCGCTGATAAGCGCGTTGAGAAAGTGTCTGACTACCTAACTGAAGGTCAAGAAGTACCAGTTAAGGTTCTTGAAATTGACCGTCAAGGCCGTGTA
>SRR27700371.35084338 A01909:184:H2CNHDSX7:1:2351:17987:13761 length=151
TACACGGCCTTGACGGTCAATTTCAAGAACCTTAACAGGTACTTCTTGACCTTCAGTTAGGTAGTCAGACACTTTCTCAACGCGCTTGTCAGCGATTTGAGAGATGTGTACTAGACCATCTTTACCTGGAAGGATAGTAACGAATGCACCG
>SRR27700371.35787228 A01909:184:H2CNHDSX7:1:2364:22724:34428 length=151
TGCTCTTTGTGAAGAAACAGGTACTACAATCGAAATCGAAGACGATGGTACAATCAAGATTGCTGCTACTGAAGGCGCAGCTGCTAAAGAAGCTATCCGTCGTATCGAAGAGATCACTGCTGAAGTTGAAGTTGGCCGCATTTACCAAGGT
>SRR27700371.37138183 A01909:184:H2CNHDSX7:1:2413:20980:14685 length=151
GGTGCATTCGTTACTATCCTTCCAGGTAAAGATGGTCTAGTACACATCTCTCAAATCGCTGACAAGCGCGTTGAGAAAGTGTCTGACTACCTAACTGAAGGTCAAGAAGTACCTGTTAAGGTTCTTGAAATTGACCGTCAAGGCCGTGTAC
>SRR27700371.37394486 A01909:184:H2CNHDSX7:1:2418:15962:13886 length=151
AAAGATGTTATCGGTAAAGGCGGCGCAGTTATCCGTGCTCTTTGTGAAGAAACAGGTACTACAATCGAAATCGAAGACGATGGCACAATCAAGATTGCTGCTACTGAAGGCGCAGCTGCTAAAGAAGCTATCCGTCGTATCGAAGAGATCA
>SRR27700371.42258832 A01909:184:H2CNHDSX7:1:2535:8639:7279 length=151
TACACGGCCTTGACGGTCAATTTCAAGAACCTTAACAGGTACTTCTTGACCTTCAGTTAGGTAGTCAGACACTTTCTCAACGCGCTTGTCAGCGATTTGAGAGATGTGTACTAGACCATCTTTACCTGGAAGGATAGTAACGAATGCACCG
>SRR27700371.43771099 A01909:184:H2CNHDSX7:1:2564:24451:21950 length=151
CGGTGCATTCGTTACTATCCTTCCAGGTAAAGATGGTCTAGTACACATCTCTCAAATCGCTGATAAGCGCGTTGAGAAAGTGTCTGACTACCTAACTGAAGGTCAAGAAGTACCAGTTAAGGTTCTTGAAATTGACCGTCAAGGCCGTGTA
>SRR27700371.43771229 A01909:184:H2CNHDSX7:1:2564:24406:22028 length=151
CGGTGCATTCGTTACTATCCTTCCAGGTAAAGATGGTCTAGTACACATCTCTCAAATCGCTGATAAGCGCGTTGAGAAAGTGTCTGACTACCTAACTGAAGGTCAAGAAGTACCAGTTAAGGGTCTTGAAATTGACCGTCAAGGCCGTGTA
>SRR27700371.44820543 A01909:184:H2CNHDSX7:1:2607:3278:15468 length=151
ACAATGAAAATCAGCTCTGATAAGATCAAAGATGTTATCGGTAAAGGCGGCGCAGTTATCCGTGCTCTTTGTGAAGAAACAGGTACTACAATCGAAATCGAAGACGATGGCACAATCAAGATTGCTGCTACTGAAGGCGCAGCTGCTAAAG
"""
        bad_seqs2 = """
        >SRR27700371.281145 A01909:184:H2CNHDSX7:1:1106:1696:19398 length=151
TACTCCTGTTATGACCGCAATAAAAAAGGGGAGCGATATGCTCCCCTTTATAACATTATTTTTTACCAATGGGTTAGAAATCACTAAAAAGAGATATTCCCATCAATACGCTAAGAATTACTCAGCAGCAGGTGCTTCGCCTTCAGCTGGT
>SRR27700371.1403435 A01909:184:H2CNHDSX7:1:1128:12427:6355 length=151
AACCTGTCAGTACAAGCAGTAAACACATACTCGCGGTTTGAAACCATTTCACAAATAATTACTCCTGTTATGACCGCAATAAAAAAGGGGAGCGATATGCTCCCCTTTATAACATGCTTTTTTACCAATGGGTTAGAAATCACTAAAAAGC
>SRR27700371.2581598 A01909:184:H2CNHDSX7:1:1150:32199:25285 length=151
ACACGGCCTTGACGGTCAATTTCAAGAACCTTAACTGGTACTTCTTGACCTTCAGTTAGGTAGTCAGACACTTTCTCAACGCGCTTATCAGCGATTTGAGAGATGTGTACTAGACCATCTTTACCTGGAAGGATAGTAACGAATGCACCGA
>SRR27700371.2829642 A01909:184:H2CNHDSX7:1:1155:14326:16188 length=151
ACACGGCCTTGACGGTCAATTTCAAGAACCTTAACTGGTACTTCTTGACCTTCAGTTAGGTAGTCAGACACTTTCTCAACGCGCTTATCAGCGATTTGAGAGATGTGTACTAGACCATCTTTACCTGGAAGGATAGTAACGAATGCACCGA
>SRR27700371.3755246 A01909:184:H2CNHDSX7:1:1173:4851:3161 length=151
TCGGTGCATTCGTTACTATCCTTCCAGGTAAAGATGGTCTAGTACACATCTCTCAAATCGCTGATAAGCGCGTTGAGAAAGTGTCTGACTACCTAACTGAAGGTCAAGAAGTACCAGTTAAGGTTCTTGAAATTGACCGTCAAGGCCGTGT
>SRR27700371.3967214 A01909:184:H2CNHDSX7:1:1177:29559:4335 length=151
GGTGCATTCGTTACTATCCTTCCAGGTAAAGATGGTCTAGTACACATCTCTCAAATCGCTGATAAGCGCGTTGAGAAAGTGTCTGACTACCTAACTGAAGGTCAAGAAGTACCAGTTAAGGTTCTTGAAATTGACCGTCAAGGCCGTGTAC
>SRR27700371.3969379 A01909:184:H2CNHDSX7:1:1177:29044:5822 length=151
GGTGCATTCGTTACTATCCTTCCAGGTAAAGATGGTCTAGTACACATCTCTCAAATCGCTGATAAGCGCGTTGAGAAAGTGTCTGACTACCTAACTGAAGGTCAAGAAGTACCAGTTAAGGTTCTTGAAATTGACCGTCAAGGCCGTGTAC
>SRR27700371.7890137 A01909:184:H2CNHDSX7:1:1273:15320:26396 length=151
TCTCTGAATTCGCTCCACGTATCCACACAATGAAAATCAGCTCTGATAAGATCAAAGATGTTATCGGTAAAGGTGGCGCAGTTATCCGTGCTCTTTGTGAAGAAACAGGTACTACAATCGAAATCGAAGACGATGGCACAATTAAGATTGC
>SRR27700371.9410083 A01909:184:H2CNHDSX7:1:1324:12246:24831 length=151
CTTGGTTGCAGCGGCACAGCCATTGGTGGATAAACCCAGCGTGAGGTGTTATCTGATGTTGTCGCACAACCTGTCAGTACAAGCAGTAAACACATACTCGCGGTTTGAAACCATTTCACAAATAATTACTCCTGTTATGACCGCAATAAAA
>SRR27700371.9594429 A01909:184:H2CNHDSX7:1:1328:26458:6605 length=151
GATCACCTAGGTGATATGGACTTTAAAGTAGCAGGTACTAACGCTGGTATCACTGCACTTCAAATGGACATCAAGATCGAAGGTATCACTAAAGAGATCATGCAAATTGCTCTTAACCAAGCGCAAGGTGCACGTAAGCACATCCTTTCTG
>SRR27700371.13011471 A01909:184:H2CNHDSX7:1:1414:29577:33113 length=151
AGCGATAGTCCCATTAATAGACTAAGAATTACTCAGCAGCAGGTGCTTCGCCTTCAGCTGGCGTTTCAACTGCTTCTTTCATGCTTAGACGTACACGGCCTTGACGGTCAATTTCAAGAACCTTAACTGGTACTTCTTGACCTTCAGTCAG
>SRR27700371.13013700 A01909:184:H2CNHDSX7:1:1414:30852:34601 length=151
AGCGATAGTCCCATTAATAGACTAAGAATTACTCAGCAGCAGGTGCTTCGCCTTCAGCTGGCGTTTCAACTGCTTCTTTCATGCTTAGACGTACACGGCCTTGACGGTCAATTTCAAGAACCTTAACTGGTACTTCTTGACCTTCAGTCAG
>SRR27700371.14045650 A01909:184:H2CNHDSX7:1:1434:15962:20055 length=151
TCTGACATCCTTGGCGACGAAGATCACCTAGGTGACATGGACTTTAAAGTAGCAGGTACTAACGCTGGTATCACTGCACTTCAAATGGACATCAAGATCGAAGGTATCACTAAAGAGATCATGCAAATTGCTCTTAACCAAGCGCAAGGTG
>SRR27700371.14048634 A01909:184:H2CNHDSX7:1:1434:14977:22106 length=151
TCTGACATCCTTGGCGACGAAGATCACCTAGGTGACATGGACTATAAAGTAGCAGGTACTAACGCTGGTATCACTGCACTTCAAATGGACATCAAGATCGAAGGTATCACTAAAGAGATCATGCAAATTGCTCTTAACCAAGCGCAAGGTG
>SRR27700371.14048653 A01909:184:H2CNHDSX7:1:1434:15022:22122 length=151
TCTGACATCCTTGGCGACGAAGATCACCTAGGTGACATGGACTTTAAAGTAGCAGGTACTAACGCTGGTATCACTGCACTTCAAATGGACATCAAGATCGAAGGTATCACTAAAGAGATCATGCAAATTGCTCTTAACCAAGCGCAAGGTG
>SRR27700371.15916112 A01909:184:H2CNHDSX7:1:1469:7374:30107 length=151
GATCACCTAGGTGATATGGACTTTAAAGTAGCAGGTACTAACGCTGGTATCACTGCACTTCAAATGGACATCAAGATCGAAGGTATCACTAAAGAGATCATGCAAATTGCTCTTAACCAAGCGCAAGGTGCACGTAAGCACATCCTTTCTG
>SRR27700371.17955650 A01909:184:H2CNHDSX7:1:1530:16170:23641 length=151
TCCGTGCTCTTTGTGAAGAAACAGGTACTACAATCGAAATCGAAGACGATGGTACAATCAAGATTGCTGCTACTGAAGGCGCAGCTGCTAAAGAAGCTATCCGTCGTATCGAAGAGATCACTGCTGAAGTTGAAGTTGGCCGCATTTACCA
>SRR27700371.18351113 A01909:184:H2CNHDSX7:1:1538:1090:5384 length=151
CGGTGCATTCGTTACTATCCTTCCAGGTAAAGATGGTCTAGTACACATCTCTCAAATCGCTGATAAGCGCGTTGAGAAAGTGTCTGTCTACCTAACTGAAGGTCAAGAAGTACCAGTTAAGGTTCTTGAAATTGACCGTCAAGGACGTGTA
>SRR27700371.18716393 A01909:184:H2CNHDSX7:1:1545:3287:2832 length=151
CGGTGCATTCGTTACTATCCTTCCAGGTAAAGATGGTCTAGTACACATCTCTCAAATCGCTGATAAGCGCGTTGAGAAAGTGTCTGACTACCTAACTGAAGGTCAAGAAGTACCAGTTAAGGTTCTTGAAATTGACCGTCAAGGCCGTGTA
>SRR27700371.19217471 A01909:184:H2CNHDSX7:1:1554:1633:21637 length=151
GAAGAAACAGGTACTACAATCGAAATCGAAGACGATGGCACAATCAAGATTGCTGCTACTGAAGGCGCAGCTGCTAAAGAAGCTATCCGTCGTATCGAAGAGATCACAGCTGAAGTTGAAGTTGGCCGCATTTACCAAGGTAAAGTTGCTC
>SRR27700371.21639151 A01909:184:H2CNHDSX7:1:1622:12861:33191 length=151
ACACGGCCTTGACGGTCAATTTCAAGAACCTTAACTGGTACTTCTTGACCTTCAGTTAGGTAGTCAGACACTTTCTCAACGCGCTTATCAGCGATTTGAGAGATGTGTACTAGACCATCTTTACCTGGAAGGATAGTAACGAATGCACCGA
>SRR27700371.27341237 A01909:184:H2CNHDSX7:1:2155:3721:25942 length=151
ATCAAGATTGCTGCTACTGAAGGCGCAGCTGCTAAAGAAGCTATCCGTCGTATCGAAGAGATCACAGCTGAAGTTGAAGTTGGCCGCATTTACCAAGGTAAAGTTGCTCGTCTAGCTGACTTCGGTGCATTCGTTACTATCCTTCCAGGTA
>SRR27700371.29912070 A01909:184:H2CNHDSX7:1:2228:22453:10254 length=151
CGGTGCATTCGTTACTATCCTTCCAGGTAAAGATGGTCTAGTACACATCTCTCAAATCGCTGACAAGCGCGTTGAGAAAGTGTCTGACTACCTAACTGAAGGTCAAGAAGTACCTGTTAAGGTTCTTGAAATTGACCGTCAAGGCCGTGTA
>SRR27700371.33707918 A01909:184:H2CNHDSX7:1:2324:15347:22435 length=151
CTTGGTTGCAGCGGCACAGCCATTGGTGGATAAACCCAGCGTGAGGTGTTATCTGATGTTGTCGCACAACCTGTCAGTACAAGCAGTAAACACATACTCGCGGTTTGAAACCATTTCACAAATAATTACTCCTGTTATGACCGCAATAAAA
>SRR27700371.33938987 A01909:184:H2CNHDSX7:1:2329:2392:3286 length=151
ATGAAAGAGATTGGTTGAAGTCGAGACGCGCTAGATCACGCAGGCCAACACTGTCGTAATAGTTACCACGTTCAAACAGCATCTTAGCTCGAACTTCATCGTTCAAATCCGGGCGCTGCAATAATTGACTAAGACGTGCCATTTGAACTTC
>SRR27700371.33939283 A01909:184:H2CNHDSX7:1:2329:2582:3490 length=151
ATGAAAGAGATTGGTTGAAGTCGAGACGCGCTAGATCACGCAGGCCAACACTGTCGTAATAGTTACCACGTTCAAACAGCATCTTAGCTCGAACTTCATCGTTCAAATCCGGGCGCTGCAATAATTGACTACGACGTGCGATTTGAACTTC
>SRR27700371.33939842 A01909:184:H2CNHDSX7:1:2329:2501:3881 length=151
ATGAAAGAGATTGGTTGAAGTCGAGACGCGCTAGATCACGCAGGCCAACACTGTCGTAATAGTTACCACGTTCAAACAGCATCTTAGCTCGAACTTCATCGTTCAAATCCGGGCGCTGCAATAATTGACTAAGACGTGCACTTTGAACTTC
>SRR27700371.34166394 A01909:184:H2CNHDSX7:1:2333:29134:18724 length=151
GAAGTCGAGACGCGCTAGATCACGCAGGCCAACACTGTCGTAATAGTTACCACGTTCAAACAGCATCTTAGCTCGAACTTCATCGTTCAAATCCGGGCGCTGCAATAATTGACTAAGACGTGCAATTTGAACTTCTTGCTGAACGCTTGGT
>SRR27700371.34173398 A01909:184:H2CNHDSX7:1:2333:31412:23703 length=151
GAAGTCGAGACGCGCTAGATCACGCAGGCCAACACTGTCGTAATAGTTACCACGTTCAAACAGCATCTTAGCTCGAACTTCATCGTTCAAATCCGGGCGCTGCAATAATTGACTAAGACGTGCAATTTGAACTACTTGCTGAACGCTTGGT
>SRR27700371.34174518 A01909:184:H2CNHDSX7:1:2333:31647:24518 length=151
GAAGTCGAGACGCGCTAGATCACGCAGGCCAACACTGTCGTAATAGTTACCACGTTCAAACAGCATCTTAGCTCGAACTTCATCGTTCAAATCCGGGCGCTGCAATAATTGACTAAGACGTGCAATTTGAACTTCTTGCTGAACGCTTGGT
>SRR27700371.35084338 A01909:184:H2CNHDSX7:1:2351:17987:13761 length=151
CACTGCGGAAGTTGAAGTTGGCCGCATTTACCAAGGTAAAGTTGCTCGTCTAGCTGACTTCGGTGCATTCGTTACTATCCTTCCAGGTAAAGATGGTCTAGTACACATCTCTCAAATCGCTGACAAGCGCGTTGAGAAAGTGTCTGACTAC
>SRR27700371.35787228 A01909:184:H2CNHDSX7:1:2364:22724:34428 length=151
ACACGGCCTTGACGGTCAATTTCAAGAACCTTAACTGGTACTTCTTGACCTTCAGTTAGGTAGTCAGACACTTTCTCAACGCGCTTATCAGCGATTTGAGAGATGTGTACTAGACCATCTTTACCTGGAAGGATAGTAACGAATGCACCGA
>SRR27700371.37138183 A01909:184:H2CNHDSX7:1:2413:20980:14685 length=151
CTGTAATAACTGACTAAGGCGTGCAATTTGAACTTCTTGTTGAACACTTGGTTGCAGTGGCACTGCCATTGGTGGATAAACCCAACGAGAGGTGTTATCTGATGTTGTCGCACAACCAGTTAGTACAAGCAGTAAACACATACTCGCGGTT
>SRR27700371.37394486 A01909:184:H2CNHDSX7:1:2418:15962:13886 length=151
ACACGGCCTTGACGGTCAATTTCAAGAACCTTAACTGGTACTTCTTGACCTTCAGTTAGGTAGTCAGACACTTTCTCAACGCGCTTATCAGCGATTTGAGAGATGTGTACTAGACCATCTTTACCTGGAAGGATAGTAACGAATGCACCGA
>SRR27700371.42258832 A01909:184:H2CNHDSX7:1:2535:8639:7279 length=151
CACTGCGGAAGTTGAAGTTGGCCGCATTTACCAAGGTAAAGTTGCTCGTCTAGCTGACTTCGGTGCATTCGTTACTATCCTTCCAGGTAAAGATGGTCTAGTACACATCTCTCAAATCGCTGACAAGCGCGTTGAGAAAGTGTCTGACTAC
>SRR27700371.43771099 A01909:184:H2CNHDSX7:1:2564:24451:21950 length=151
AATCAAGCAGGGCTTGCTCTGGCGTGAGTTCATGCTGAATGATGTAAAGCCACAGAGCGCTAAACGGATCGATAGGATCATAGGTGTAGTGCTTCATCATCTCTTCATTGGCTAAGTCATAGAGCTCGCCGTAGTAGAGTGCGATAGCGCG
>SRR27700371.43771229 A01909:184:H2CNHDSX7:1:2564:24406:22028 length=151
AATCAAGCAGGGCTTGCTCTGGCGTGAGTTCATGCTGAATGATGTAAAGCCACAGAGAGCTAAACGGATCTCTAGGATCATCGGTGTAGTGCTTCATCATCTCTTCATTGGCTAAGTCATAGCGCTCGCCGTAGTAGAGTGAGATAGAGCG
>SRR27700371.44820543 A01909:184:H2CNHDSX7:1:2607:3278:15468 length=151
ACACGGCCTTGACGGTCAATTTCAAGAACCTTAACTGGTACTTCTTGACCTTCAGTTAGGTAGTCAGACACTTTCTCAACGCGCTTATCAGCGATTTGAGAGATGTGTACTAGACCATCTTTACCTGGAAGGATAGTAACGAATGCACCGA
        """
        with tempfile.NamedTemporaryFile(mode='w+t') as bad_fasta1, \
             tempfile.NamedTemporaryFile(mode='w+t') as bad_fasta2:
            bad_fasta1.write(bad_seqs1.strip())
            bad_fasta1.flush()
            bad_fasta2.write(bad_seqs2.strip())
            bad_fasta2.flush()

            # This should not raise an exception. Run singlem, outputing a taxonomic profile
            cmd = "singlem pipe -1 {} -2 {} -p /dev/stdout".format(bad_fasta1.name, bad_fasta2.name)
            stdout = extern.run(cmd)
            self.assertEqual("sample\tcoverage\ttaxonomy\n", stdout)
            


if __name__ == "__main__":
    import logging
    # logging.basicConfig(level=logging.DEBUG, format='%(asctime)s %(levelname)s: %(message)s', datefmt='%m/%d/%Y %I:%M:%S %p')
    unittest.main()