<!doctype html>

<html lang="en">

<head>
    <meta charset="utf-8">

    <title>README</title>
    <meta name="description" content="Documentation for SingleM">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <link
        href="https://fonts.googleapis.com/css2?family=Inter:wght@400;600;700&family=Source+Sans+Pro:ital,wght@0,400;0,600;0,700;1,400;1,600;1,700&display=swap"
        rel="stylesheet">

    <link rel="stylesheet" type="text/css" href="/singlem/assets/normalize.css?v=1771818232"
        media="screen" />
    <link rel="stylesheet" type="text/css" href="/singlem/assets/doctave-style.css?v=1771818232"
        media="screen" />
    <link rel="stylesheet" type="text/css" href="/singlem/assets/katex.css?v=1771818232" media="screen" />
    <link rel="stylesheet" type="text/css" href="/singlem/assets/prism-ghcolors.css?v=1771818232"
        media="screen" />

    <script>
        var DOCTAVE_TIMESTAMP = "1771818232";
        var BASE_PATH = "/singlem/";
        var color = localStorage.getItem('doctave-color')

        if (color === 'dark') {
            document.getElementsByTagName('html')[0].classList.remove('light');
            document.getElementsByTagName('html')[0].classList.add('dark');
        } else {
            document.getElementsByTagName('html')[0].classList.remove('dark');
            document.getElementsByTagName('html')[0].classList.add('light');
        }
    </script>

    
</head>

<body>
    <label for='menu-toggle-switch' class='menu-toggle-button'>
        â˜°
    </label>
    <input type="checkbox" id="menu-toggle-switch" value='0' />
    <div class='page'>
        <div class='header'>
            <div class='logo'>
                
                <a href="/singlem/">
                    <img src="/singlem/sandpiper.png" alt='SingleM logo'></img>
                </a>
                
                <h2 class='project-name'>
                    <a href="/singlem/">
                        SingleM
                    </a>
                </h2>
            </div>
            <div class='search'>
                <form id='search-form'>
    <input type='text' id='search-box' autocomplete="off" placeholder="Search..."></input>
    <span class='search-icon'>S</span>
    <ul id='search-results'></ul>
</form>

            </div>
            <div class='header-dummy-right'>
            </div>
        </div>
        <div class='container'>
            <div class='sidebar-left'>
                <nav class='site-nav'>
    <ul>
        
            <li><a href="/singlem/Lyrebird">Lyrebird (phage profiling)</a></li>
            
        
            <li><a href="/singlem/Installation">Installation</a></li>
            
        
            <li><a href="/singlem/Glossary">Glossary</a></li>
            
        
            <li><a href="/singlem/FAQ">FAQ</a></li>
            
        
            <li><a href="/singlem/tools">Usage</a></li>
            
                <ul>
    
        <li><a href="/singlem/tools/data">SingleM data</a></li>
        
    
        <li><a href="/singlem/tools/pipe">SingleM pipe</a></li>
        
    
        <li><a href="/singlem/tools/summarise">SingleM summarise</a></li>
        
    
        <li><a href="/singlem/tools/renew">SingleM renew</a></li>
        
    
        <li><a href="/singlem/tools/supplement">SingleM supplement</a></li>
        
    
        <li><a href="/singlem/tools/prokaryotic_fraction">SingleM prokaryotic_fraction</a></li>
        
    
        <li><a href="/singlem/tools/appraise">SingleM appraise</a></li>
        
    
        <li><a href="/singlem/tools/lyrebird_data">Lyrebird data</a></li>
        
    
        <li><a href="/singlem/tools/lyrebird_pipe">Lyrebird pipe</a></li>
        
    
</ul>

            
        
            <li><a href="/singlem/advanced">Advanced modes</a></li>
            
                <ul>
    
        <li><a href="/singlem/advanced/query">SingleM query</a></li>
        
    
        <li><a href="/singlem/advanced/makedb">SingleM makedb</a></li>
        
    
        <li><a href="/singlem/advanced/condense">SingleM condense</a></li>
        
    
        <li><a href="/singlem/advanced/seqs">SingleM seqs</a></li>
        
    
        <li><a href="/singlem/advanced/create">SingleM create</a></li>
        
    
        <li><a href="/singlem/advanced/metapackage">SingleM metapackage</a></li>
        
    
        <li><a href="/singlem/advanced/regenerate">SingleM regenerate</a></li>
        
    
        <li><a href="/singlem/advanced/lyrebird_condense">Lyrebird condense</a></li>
        
    
        <li><a href="/singlem/advanced/lyrebird_renew">Lyrebird renew</a></li>
        
    
</ul>

            
        
    </ul>
    <span id='light-dark-mode-switch'>
        <svg xmlns="http://www.w3.org/2000/svg" id="dark-mode-icon" viewBox="0 0 20 20" fill="currentColor">
            <path d="M17.293 13.293A8 8 0 016.707 2.707a8.001 8.001 0 1010.586 10.586z" />
        </svg>

        <svg xmlns="http://www.w3.org/2000/svg" id="light-mode-icon" viewBox="0 0 20 20" fill="currentColor">
            <path fill-rule="evenodd" d="M10 2a1 1 0 011 1v1a1 1 0 11-2 0V3a1 1 0 011-1zm4 8a4 4 0 11-8 0 4 4 0 018 0zm-.464 4.95l.707.707a1 1 0 001.414-1.414l-.707-.707a1 1 0 00-1.414 1.414zm2.12-10.607a1 1 0 010 1.414l-.706.707a1 1 0 11-1.414-1.414l.707-.707a1 1 0 011.414 0zM17 11a1 1 0 100-2h-1a1 1 0 100 2h1zm-7 4a1 1 0 011 1v1a1 1 0 11-2 0v-1a1 1 0 011-1zM5.05 6.464A1 1 0 106.465 5.05l-.708-.707a1 1 0 00-1.414 1.414l.707.707zm1.414 8.486l-.707.707a1 1 0 01-1.414-1.414l.707-.707a1 1 0 011.414 1.414zM4 11a1 1 0 100-2H3a1 1 0 000 2h1z" clip-rule="evenodd" />
        </svg>
    </span>
</nav>

            </div>
            <div class='doctave-content'>
                
<p><a href="https://sandpiper.qut.edu.au"><img src="./sandpiper_small.png" alt="Sandpiper logo"></a>
<a href="https://aroneys.github.io/binchicken"><img src="./binchicken_small.png" alt="Bin Chicken logo"></a>
<a href="/singlem/Lyrebird"><img src="./lyrebird_small.png" alt="Lyrebird logo"></a></p>
<p><a href="https://github.com/wwood/singlem/actions"><img src="https://github.com/wwood/singlem/actions/workflows/test-singlem.yml/badge.svg" alt="Current Build"></a>
<a href="https://anaconda.org/bioconda/singlem"><img src="https://img.shields.io/conda/vn/bioconda/singlem?label=Conda&amp;color=43b02a" alt="Conda version"></a>
<a href="https://anaconda.org/bioconda/singlem"><img src="https://img.shields.io/conda/dn/bioconda/singlem?label=Downloads&amp;color=43b02a" alt="Conda downloads"></a>
<a href="https://hub.docker.com/r/wwood/singlem/tags"><img src="https://img.shields.io/docker/v/wwood/singlem?label=Docker&amp;color=1D63ED" alt="Docker version"></a>
<a href="https://hub.docker.com/r/wwood/singlem"><img src="https://img.shields.io/docker/pulls/wwood/singlem.svg?label=Pulls&amp;color=1D63ED" alt="Docker pulls"></a>
<a href="https://pypi.org/project/singlem/"><img src="https://img.shields.io/pypi/v/singlem.svg?label=PyPI&amp;color=ffd43b" alt="PyPI version"></a></p>
<p>Welcome.</p>
<p>At heart, SingleM is a tool for profiling shotgun (both short and long-read) metagenomes. It <a href="https://doi.org/10.1038/s41587-025-02738-1">shows</a> good accuracy in estimating the relative abundances of community members, and has a particular strength in dealing with novel lineages.</p>
<p>It was originally designed to determine the relative abundance of bacterial and archaeal taxa in a sample. Microbial SingleM has been applied to ~700,000 public metagenomes. The resulting data are available at the <a href="https://sandpiper.qut.edu.au">Sandpiper companion website</a>.</p>
<p>Recent versions have added features:</p>
<ul>
<li>Long-read input support (v0.20.0). Either Nanopore &gt;= R10.4.1 or PacBio HiFi reads are recommended to ensure reliable taxonomic profiling.</li>
<li>Profiling of dsDNA phages (v0.19.0, updated DB in v0.20.0). See <a href="/singlem/Lyrebird">Lyrebird</a>.</li>
</ul>
<p>The method it uses also it suitable for some related tasks, such as assessing eukaryotic contamination, finding bias in genome recovery, and lineage-targeted MAG recovery. It can also be used as the basis for choosing metagenomes which, when coassembled, maximise the recovery of novel MAGs (see <a href="https://aroneys.github.io/binchicken/">Bin Chicken</a>).</p>
<p>The main idea of SingleM is to profile metagenomes by targeting short 20 amino acid stretches ("<em>windows</em>") within single copy marker genes. It finds reads which cover an entire window, and analyses these further. By constraining analysis to these short windows, it becomes possible to know how novel each read is compared to known genomes. Then, using the fact that each analysed gene is (almost always) found exactly once in each genome, the abundance of each lineage can be accurately estimated.</p>
<p>There are several tools (subcommands) which can be used after <a href="/singlem/Installation">installation</a>:</p>
<ul>
<li><a href="/singlem/tools/pipe">singlem pipe</a> - the main workflow which generates OTU tables and <a href="https://gtdb.ecogenomic.org/">GTDB</a> taxonomic profiles. </li>
<li><a href="/singlem/tools/summarise">single summarise</a> - Mechanical transformations of <code>singlem pipe</code> results.</li>
<li><a href="/singlem/tools/renew">singlem renew</a> - Given previously generated results, re-run the pipeline with a new reference sequence/taxonomy database.</li>
<li><a href="/singlem/tools/supplement">singlem supplement</a> - Add new genomes to a reference metapackage.</li>
<li><a href="/singlem/tools/prokaryotic_fraction">singlem prokaryotic_fraction</a> - How much of a metagenome is prokaryotic? (also available as <code>microbial_fraction</code>, the deprecated name)</li>
<li><a href="/singlem/tools/appraise">singlem appraise</a> - How much of a metagenome do the genomes or assembly represent?</li>
</ul>
<p>And more specialised / expert modes:</p>
<ul>
<li><a href="/singlem/advanced/condense">singlem condense</a> - Given an OTU table, summarise the results into a taxonomic profile.</li>
<li><a href="/singlem/advanced/makedb">singlem makedb</a> &amp; <a href="/singlem/advanced/query">query</a>- Create a database of OTU sequences and query it using various sequence similarity methods e.g. <a href="https://github.com/wwood/smafa">smafa</a>.</li>
</ul>
<h2 id="help">Help</h2>
<p>If you have any questions or comments, raise a <a href="https://github.com/wwood/singlem/issues">GitHib issue</a> or just send us an <a href="https://research.qut.edu.au/cmr/team/ben-woodcroft/">email</a>.</p>
<h2 id="how-to-use-singlem-with-ai-assistants-(claude/chatgpt/gemini-etc.)">How to use SingleM with AI assistants (Claude/ChatGPT/Gemini etc.)</h2>
<p>There is a <a href="https://raw.githubusercontent.com/wwood/singlem/main/docs/SKILL.md">SKILL.md</a> file that is designed to be used for teaching AI assistants about SingleM's usage. </p>
<h2 id="license">License</h2>
<p>SingleM is developed by the <a href="https://research.qut.edu.au/cmr/team/ben-woodcroft/">Woodcroft lab</a> at the <a href="https://research.qut.edu.au/cmr">Centre for Microbiome Research</a>, School of Biomedical Sciences, QUT, with contributions from many helpful people including <a href="https://github.com/AroneyS">Samuel Aroney</a>, <a href="https://github.com/rzhao-2">Rossen Zhao</a>, and <a href="https://github.com/EisenRa">Raphael Eisenhofer</a>. It is licensed under <a href="https://gnu.org/licenses/gpl.html">GPL3 or later</a>.</p>
<p>The source code is available at <a href="https://github.com/wwood/singlem">https://github.com/wwood/singlem</a>.</p>
<h2 id="citations">Citations</h2>

<h3 id="profiling-microbial-communities-with-singlem-/-sandpiper">Profiling microbial communities with SingleM / Sandpiper</h3>
<p>Ben J. Woodcroft, Samuel T. N. Aroney, Rossen Zhao, Mitchell Cunningham, Joshua A. M. Mitchell, Rizky Nurdiansyah, Linda Blackall &amp; Gene W. Tyson. <em>Comprehensive taxonomic identification of microbial species in metagenomic data using SingleM and Sandpiper.</em> Nat Biotechnol (2025). <a href="https://doi.org/10.1038/s41587-025-02738-1">https://doi.org/10.1038/s41587-025-02738-1</a>.</p>
<h3 id="singlem-prokaryotic_fraction">SingleM prokaryotic_fraction</h3>
<p>Raphael Eisenhofer, Antton Alberdi, Ben J. Woodcroft, 2024. <em>Large-scale estimation of bacterial and archaeal DNA prevalence in metagenomes reveals biome-specific patterns.</em> bioRxiv, pp.2024-05; <a href="https://doi.org/10.1101/2024.05.16.594470">https://doi.org/10.1101/2024.05.16.594470</a>.</p>
<h3 id="singlem-powered-coassembly-with-bin-chicken">SingleM-powered coassembly with Bin Chicken</h3>
<p>Samuel T. N. Aroney, Rhys J. Newell, Gene W. Tyson and Ben J. Woodcroft, 2024. <em>Bin Chicken: targeted metagenomic coassembly for the efficient recovery of novel genomes.</em> bioRxiv, pp.2024-11. <a href="https://doi.org/10.1101/2024.11.24.625082">https://doi.org/10.1101/2024.11.24.625082</a>.</p>
<h3 id="lyrebird">Lyrebird</h3>
<p>Rossen Zhao, Gene W. Tyson, Ben J. Woodcroft. <em>Lyrebird: a tool for profiling dsDNA phage communities in metagenomic data.</em> (in preparation).</p>

            </div>
            <div class='sidebar-right'>
                <div class='page-nav' id='page-nav'>
                    <p class='page-nav-header'>On this page</p>
                    <ul>
                        
                        <li class='page-nav-level-2'>
                            <a href='#help'>Help</a>
                        </li>
                        
                        <li class='page-nav-level-2'>
                            <a href='#how-to-use-singlem-with-ai-assistants-(claude/chatgpt/gemini-etc.)'>How to use SingleM with AI assistants (Claude/ChatGPT/Gemini etc.)</a>
                        </li>
                        
                        <li class='page-nav-level-2'>
                            <a href='#license'>License</a>
                        </li>
                        
                        <li class='page-nav-level-2'>
                            <a href='#citations'>Citations</a>
                        </li>
                        
                        <li class='page-nav-level-3'>
                            <a href='#profiling-microbial-communities-with-singlem-/-sandpiper'>Profiling microbial communities with SingleM / Sandpiper</a>
                        </li>
                        
                        <li class='page-nav-level-3'>
                            <a href='#singlem-prokaryotic_fraction'>SingleM prokaryotic_fraction</a>
                        </li>
                        
                        <li class='page-nav-level-3'>
                            <a href='#singlem-powered-coassembly-with-bin-chicken'>SingleM-powered coassembly with Bin Chicken</a>
                        </li>
                        
                        <li class='page-nav-level-3'>
                            <a href='#lyrebird'>Lyrebird</a>
                        </li>
                        
                    </ul>
                </div>
            </div>
            <div class='wave-container'>
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1440 320">
                    <path fill-opacity="0.35"
                        d="M0,192L60,213.3C120,235,240,277,360,277.3C480,277,600,235,720,192C840,149,960,107,1080,122.7C1200,139,1320,213,1380,250.7L1440,288L1440,320L1380,320C1320,320,1200,320,1080,320C960,320,840,320,720,320C600,320,480,320,360,320C240,320,120,320,60,320L0,320Z">
                    </path>
                    <path fill-opacity="0.5"
                        d="M0,160L60,181.3C120,203,240,245,360,229.3C480,213,600,139,720,138.7C840,139,960,213,1080,229.3C1200,245,1320,203,1380,181.3L1440,160L1440,320L1380,320C1320,320,1200,320,1080,320C960,320,840,320,720,320C600,320,480,320,360,320C240,320,120,320,60,320L0,320Z">
                    </path>
                    <path fill-opacity="0.2"
                        d="M0,224L60,197.3C120,171,240,117,360,117.3C480,117,600,171,720,186.7C840,203,960,181,1080,186.7C1200,192,1320,224,1380,240L1440,256L1440,320L1380,320C1320,320,1200,320,1080,320C960,320,840,320,720,320C600,320,480,320,360,320C240,320,120,320,60,320L0,320Z">
                    </path>
                </svg>
                <p>Powered by <a href='https://cli.doctave.com' target='_blank'>Doctave</a></p>
            </div>
        </div>
    </div>
    <script type="text/javascript" src="/singlem/assets/katex.js?v=1771818232"></script>
    <script type="text/javascript" src="/singlem/assets/mermaid.js?v=1771818232"></script>
    <script type="text/javascript" src="/singlem/assets/elasticlunr.js?v=1771818232"></script>
    <script type="text/javascript" src="/singlem/assets/prism.js?v=1771818232"></script>
    <script type="text/javascript" src="/singlem/assets/doctave-app.js?v=1771818232"></script>

    
</body>

</html>