<!doctype html>

<html lang="en">

<head>
    <meta charset="utf-8">

    <title>SingleM supplement</title>
    <meta name="description" content="Documentation for SingleM">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <link
        href="https://fonts.googleapis.com/css2?family=Inter:wght@400;600;700&family=Source+Sans+Pro:ital,wght@0,400;0,600;0,700;1,400;1,600;1,700&display=swap"
        rel="stylesheet">

    <link rel="stylesheet" type="text/css" href="/singlem/assets/normalize.css?v=1771818232"
        media="screen" />
    <link rel="stylesheet" type="text/css" href="/singlem/assets/doctave-style.css?v=1771818232"
        media="screen" />
    <link rel="stylesheet" type="text/css" href="/singlem/assets/katex.css?v=1771818232" media="screen" />
    <link rel="stylesheet" type="text/css" href="/singlem/assets/prism-ghcolors.css?v=1771818232"
        media="screen" />

    <script>
        var DOCTAVE_TIMESTAMP = "1771818232";
        var BASE_PATH = "/singlem/";
        var color = localStorage.getItem('doctave-color')

        if (color === 'dark') {
            document.getElementsByTagName('html')[0].classList.remove('light');
            document.getElementsByTagName('html')[0].classList.add('dark');
        } else {
            document.getElementsByTagName('html')[0].classList.remove('dark');
            document.getElementsByTagName('html')[0].classList.add('light');
        }
    </script>

    
</head>

<body>
    <label for='menu-toggle-switch' class='menu-toggle-button'>
        â˜°
    </label>
    <input type="checkbox" id="menu-toggle-switch" value='0' />
    <div class='page'>
        <div class='header'>
            <div class='logo'>
                
                <a href="/singlem/">
                    <img src="/singlem/sandpiper.png" alt='SingleM logo'></img>
                </a>
                
                <h2 class='project-name'>
                    <a href="/singlem/">
                        SingleM
                    </a>
                </h2>
            </div>
            <div class='search'>
                <form id='search-form'>
    <input type='text' id='search-box' autocomplete="off" placeholder="Search..."></input>
    <span class='search-icon'>S</span>
    <ul id='search-results'></ul>
</form>

            </div>
            <div class='header-dummy-right'>
            </div>
        </div>
        <div class='container'>
            <div class='sidebar-left'>
                <nav class='site-nav'>
    <ul>
        
            <li><a href="/singlem/Lyrebird">Lyrebird (phage profiling)</a></li>
            
        
            <li><a href="/singlem/Installation">Installation</a></li>
            
        
            <li><a href="/singlem/Glossary">Glossary</a></li>
            
        
            <li><a href="/singlem/FAQ">FAQ</a></li>
            
        
            <li><a href="/singlem/tools">Usage</a></li>
            
                <ul>
    
        <li><a href="/singlem/tools/data">SingleM data</a></li>
        
    
        <li><a href="/singlem/tools/pipe">SingleM pipe</a></li>
        
    
        <li><a href="/singlem/tools/summarise">SingleM summarise</a></li>
        
    
        <li><a href="/singlem/tools/renew">SingleM renew</a></li>
        
    
        <li><a class="active" href="/singlem/tools/supplement">SingleM supplement</a></li>
        
    
        <li><a href="/singlem/tools/prokaryotic_fraction">SingleM prokaryotic_fraction</a></li>
        
    
        <li><a href="/singlem/tools/appraise">SingleM appraise</a></li>
        
    
        <li><a href="/singlem/tools/lyrebird_data">Lyrebird data</a></li>
        
    
        <li><a href="/singlem/tools/lyrebird_pipe">Lyrebird pipe</a></li>
        
    
</ul>

            
        
            <li><a href="/singlem/advanced">Advanced modes</a></li>
            
                <ul>
    
        <li><a href="/singlem/advanced/query">SingleM query</a></li>
        
    
        <li><a href="/singlem/advanced/makedb">SingleM makedb</a></li>
        
    
        <li><a href="/singlem/advanced/condense">SingleM condense</a></li>
        
    
        <li><a href="/singlem/advanced/seqs">SingleM seqs</a></li>
        
    
        <li><a href="/singlem/advanced/create">SingleM create</a></li>
        
    
        <li><a href="/singlem/advanced/metapackage">SingleM metapackage</a></li>
        
    
        <li><a href="/singlem/advanced/regenerate">SingleM regenerate</a></li>
        
    
        <li><a href="/singlem/advanced/lyrebird_condense">Lyrebird condense</a></li>
        
    
        <li><a href="/singlem/advanced/lyrebird_renew">Lyrebird renew</a></li>
        
    
</ul>

            
        
    </ul>
    <span id='light-dark-mode-switch'>
        <svg xmlns="http://www.w3.org/2000/svg" id="dark-mode-icon" viewBox="0 0 20 20" fill="currentColor">
            <path d="M17.293 13.293A8 8 0 016.707 2.707a8.001 8.001 0 1010.586 10.586z" />
        </svg>

        <svg xmlns="http://www.w3.org/2000/svg" id="light-mode-icon" viewBox="0 0 20 20" fill="currentColor">
            <path fill-rule="evenodd" d="M10 2a1 1 0 011 1v1a1 1 0 11-2 0V3a1 1 0 011-1zm4 8a4 4 0 11-8 0 4 4 0 018 0zm-.464 4.95l.707.707a1 1 0 001.414-1.414l-.707-.707a1 1 0 00-1.414 1.414zm2.12-10.607a1 1 0 010 1.414l-.706.707a1 1 0 11-1.414-1.414l.707-.707a1 1 0 011.414 0zM17 11a1 1 0 100-2h-1a1 1 0 100 2h1zm-7 4a1 1 0 011 1v1a1 1 0 11-2 0v-1a1 1 0 011-1zM5.05 6.464A1 1 0 106.465 5.05l-.708-.707a1 1 0 00-1.414 1.414l.707.707zm1.414 8.486l-.707.707a1 1 0 01-1.414-1.414l.707-.707a1 1 0 011.414 1.414zM4 11a1 1 0 100-2H3a1 1 0 000 2h1z" clip-rule="evenodd" />
        </svg>
    </span>
</nav>

            </div>
            <div class='doctave-content'>
                <h1 id="singlem-supplement">singlem supplement</h1>
<p>The SingleM <code>supplement</code> subcommand adds genomes to a SingleM metapackage.</p>
<p><strong>TLDR</strong>: Supplement a metapackage with new genomes like so:</p>
<pre><code>singlem supplement --new-genome-fasta-files &lt;genome1.fna&gt; &lt;genome2.fna&gt; \
    --output-metapackage &lt;supplemented.smpkg&gt;
</code></pre>
<h2 id="gtdb-tk">GTDB-Tk</h2>
<p>In order to add genomes to a metapackage, their taxonomy is required. SingleM <code>supplement</code> can generate this taxonomy for new genomes using GTDB-Tk. However, GTDB-Tk is not installed by default, and so it must be installed separately.</p>
<p>For instance, if you are using conda to manage dependencies, use:</p>
<pre><code>conda install gtdbtk
</code></pre>
<p>Note that the version of GTDB-Tk installed must match the GTDB release of the metapackage. The <a href="https://ecogenomics.github.io/GTDBTk/installing/index.html">GTDB-Tk documentation</a> provides guidance on installation and compatibility.</p>
<h1 id="options">OPTIONS</h1>
<p><strong>--new-genome-fasta-files</strong> <em>NEW_GENOME_FASTA_FILES</em> [<em>NEW_GENOME_FASTA_FILES</em> ...]</p>
<p>FASTA files of new genomes</p>
<p><strong>--new-genome-fasta-files-list</strong> <em>NEW_GENOME_FASTA_FILES_LIST</em> [<em>NEW_GENOME_FASTA_FILES_LIST</em> ...]</p>
<p>File containing FASTA file paths of new genomes</p>
<p><strong>--input-metapackage</strong> <em>INPUT_METAPACKAGE</em></p>
<p>metapackage to build upon [default: Use default package]</p>
<p><strong>--output-metapackage</strong> <em>OUTPUT_METAPACKAGE</em></p>
<p>output metapackage</p>
<p><strong>--threads</strong> <em>THREADS</em></p>
<p>parallelisation</p>
<h1 id="taxonomy">TAXONOMY</h1>
<p><strong>--new-fully-defined-taxonomies</strong> <em>NEW_FULLY_DEFINED_TAXONOMIES</em></p>
<p>newline separated file containing taxonomies of new genomes
(path&lt;TAB&gt;taxonomy). Must be fully specified to species level. If
not specified, the taxonomy will be inferred from the new genomes
using GTDB-tk or read from --taxonomy-file [default: not set, run
GTDBtk].</p>
<p><strong>--taxonomy-file</strong> <em>TAXONOMY_FILE</em></p>
<p>A 2 column tab-separated file containing each genome's taxonomy as
output by GTDBtk [default: not set, run GTDBtk]</p>
<p><strong>--gtdbtk-output-directory</strong> <em>GTDBTK_OUTPUT_DIRECTORY</em></p>
<p>use this GTDBtk result. Not used if --new-taxonomies is used
[default: not set, run GTDBtk]</p>
<p><strong>--pplacer-threads</strong> <em>PPLACER_THREADS</em></p>
<p>for GTDBtk classify_wf</p>
<p><strong>--output-taxonomies</strong> <em>OUTPUT_TAXONOMIES</em></p>
<p>TSV output file of taxonomies of new genomes, whether they are novel
species or not.</p>
<p><strong>--skip-taxonomy-check</strong></p>
<p>skip check which ensures that GTDBtk assigned taxonomies are
concordant with the old metapackage's [default: do the check]</p>
<h1 id="quality-filtering-of-new-genomes">QUALITY FILTERING OF NEW GENOMES</h1>
<p><strong>--checkm2-quality-file</strong> <em>CHECKM2_QUALITY_FILE</em></p>
<p>CheckM2 quality file of new genomes</p>
<p><strong>--no-quality-filter</strong></p>
<p>skip quality filtering</p>
<p><strong>--checkm2-min-completeness</strong> <em>CHECKM2_MIN_COMPLETENESS</em></p>
<p>minimum completeness for CheckM2 [default: 50]</p>
<p><strong>--checkm2-max-contamination</strong> <em>CHECKM2_MAX_CONTAMINATION</em></p>
<p>maximum contamination for CheckM2 [default: 10]</p>
<h1 id="dereplication">DEREPLICATION</h1>
<p><strong>--no-dereplication</strong></p>
<p>Assume genome inputs are already dereplicated</p>
<p><strong>--dereplicate-with-galah</strong></p>
<p>Run galah to dereplicate genomes at species level</p>
<h1 id="less-common-options">LESS COMMON OPTIONS</h1>
<p><strong>--hmmsearch-evalue</strong> <em>HMMSEARCH_EVALUE</em></p>
<p>evalue for hmmsearch run on proteins to gather markers [default:
1e-20]</p>
<p><strong>--gene-definitions</strong> <em>GENE_DEFINITIONS</em></p>
<p>Tab-separated file of
genome_fasta&lt;TAB&gt;transcript_fasta&lt;TAB&gt;protein_fasta [default:
undefined, call genes using Prodigal]</p>
<p><strong>--working-directory</strong> <em>WORKING_DIRECTORY</em></p>
<p>working directory [default: use a temporary directory]</p>
<p><strong>--no-taxon-genome-lengths</strong></p>
<p>Do not include taxon genome lengths in updated metapackage</p>
<p><strong>--ignore-taxonomy-database-incompatibility</strong></p>
<p>Do not halt when the old metapackage is not the default metapackage.</p>
<p><strong>--new-taxonomy-database-name</strong> <em>NEW_TAXONOMY_DATABASE_NAME</em></p>
<p>Name of the taxonomy database to record in the created metapackage
[default: custom_taxonomy_database]</p>
<p><strong>--new-taxonomy-database-version</strong> <em>NEW_TAXONOMY_DATABASE_VERSION</em></p>
<p>Version of the taxonomy database to use [default: None]</p>
<h1 id="other-general-options">OTHER GENERAL OPTIONS</h1>
<p><strong>--debug</strong></p>
<p>output debug information</p>
<p><strong>--version</strong></p>
<p>output version information and quit</p>
<p><strong>--quiet</strong></p>
<p>only output errors</p>
<p><strong>--full-help</strong></p>
<p>print longer help message</p>
<p><strong>--full-help-roff</strong></p>
<p>print longer help message in ROFF (manpage) format</p>
<h1 id="authors">AUTHORS</h1>
<blockquote>
<pre><code>Ben J. Woodcroft, Centre for Microbiome Research, School of Biomedical Sciences, Faculty of Health, Queensland University of Technology
Samuel Aroney, Centre for Microbiome Research, School of Biomedical Sciences, Faculty of Health, Queensland University of Technology
Raphael Eisenhofer, Centre for Evolutionary Hologenomics, University of Copenhagen, Denmark
Rossen Zhao, Centre for Microbiome Research, School of Biomedical Sciences, Faculty of Health, Queensland University of Technology
</code></pre>
</blockquote>

            </div>
            <div class='sidebar-right'>
                <div class='page-nav' id='page-nav'>
                    <p class='page-nav-header'>On this page</p>
                    <ul>
                        
                        <li class='page-nav-level-1'>
                            <a href='#singlem-supplement'>singlem supplement</a>
                        </li>
                        
                        <li class='page-nav-level-2'>
                            <a href='#gtdb-tk'>GTDB-Tk</a>
                        </li>
                        
                        <li class='page-nav-level-1'>
                            <a href='#options'>OPTIONS</a>
                        </li>
                        
                        <li class='page-nav-level-1'>
                            <a href='#taxonomy'>TAXONOMY</a>
                        </li>
                        
                        <li class='page-nav-level-1'>
                            <a href='#quality-filtering-of-new-genomes'>QUALITY FILTERING OF NEW GENOMES</a>
                        </li>
                        
                        <li class='page-nav-level-1'>
                            <a href='#dereplication'>DEREPLICATION</a>
                        </li>
                        
                        <li class='page-nav-level-1'>
                            <a href='#less-common-options'>LESS COMMON OPTIONS</a>
                        </li>
                        
                        <li class='page-nav-level-1'>
                            <a href='#other-general-options'>OTHER GENERAL OPTIONS</a>
                        </li>
                        
                        <li class='page-nav-level-1'>
                            <a href='#authors'>AUTHORS</a>
                        </li>
                        
                    </ul>
                </div>
            </div>
            <div class='wave-container'>
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1440 320">
                    <path fill-opacity="0.35"
                        d="M0,192L60,213.3C120,235,240,277,360,277.3C480,277,600,235,720,192C840,149,960,107,1080,122.7C1200,139,1320,213,1380,250.7L1440,288L1440,320L1380,320C1320,320,1200,320,1080,320C960,320,840,320,720,320C600,320,480,320,360,320C240,320,120,320,60,320L0,320Z">
                    </path>
                    <path fill-opacity="0.5"
                        d="M0,160L60,181.3C120,203,240,245,360,229.3C480,213,600,139,720,138.7C840,139,960,213,1080,229.3C1200,245,1320,203,1380,181.3L1440,160L1440,320L1380,320C1320,320,1200,320,1080,320C960,320,840,320,720,320C600,320,480,320,360,320C240,320,120,320,60,320L0,320Z">
                    </path>
                    <path fill-opacity="0.2"
                        d="M0,224L60,197.3C120,171,240,117,360,117.3C480,117,600,171,720,186.7C840,203,960,181,1080,186.7C1200,192,1320,224,1380,240L1440,256L1440,320L1380,320C1320,320,1200,320,1080,320C960,320,840,320,720,320C600,320,480,320,360,320C240,320,120,320,60,320L0,320Z">
                    </path>
                </svg>
                <p>Powered by <a href='https://cli.doctave.com' target='_blank'>Doctave</a></p>
            </div>
        </div>
    </div>
    <script type="text/javascript" src="/singlem/assets/katex.js?v=1771818232"></script>
    <script type="text/javascript" src="/singlem/assets/mermaid.js?v=1771818232"></script>
    <script type="text/javascript" src="/singlem/assets/elasticlunr.js?v=1771818232"></script>
    <script type="text/javascript" src="/singlem/assets/prism.js?v=1771818232"></script>
    <script type="text/javascript" src="/singlem/assets/doctave-app.js?v=1771818232"></script>

    
</body>

</html>